program/src/processor.rs

```diff
@@ -601,83 +658,127 @@ pub fn process_withdraw_inactive_stake(
         .validators
         .get_mut(accounts.validator_vote_account.key)?;
 
-    let mut observed_total = Lamports(0);
+    let mut stake_observed_total = Lamports(0);
     let mut excess_removed = Lamports(0);
-    let mut stake_accounts = accounts.stake_accounts.iter();
-    let begin = validator.entry.stake_seeds.begin;
-    let end = validator.entry.stake_seeds.end;
+    let n_stake_accounts = validator.entry.stake_seeds.end - validator.entry.stake_seeds.begin;
+    let n_unstake_accounts =
+        validator.entry.unstake_seeds.end - validator.entry.unstake_seeds.begin;
+
+    if accounts.stake_accounts.len() as u64 != n_stake_accounts + n_unstake_accounts {
+        msg!("Wrong number of stake accounts provided, expected {} stake accounts and {} unstake accounts, \
+            but got {} accounts.", n_stake_accounts, n_unstake_accounts, accounts.stake_accounts.len());
+        return Err(LidoError::InvalidStakeAccount.into());
+    }
+    // Does not panic, because len = n_stake_accounts + n_unstake_accounts >= n_stake_accounts.
+    let (stake_accounts, unstake_accounts) =
+        accounts.stake_accounts.split_at(n_stake_accounts as usize);
 
     // Visit the stake accounts one by one, and check how much SOL is in there.
-    for seed in &validator.entry.stake_seeds {
+    for (i, seed) in validator.entry.stake_seeds.into_iter().enumerate() {
         let (stake_account_address, _bump_seed) =
             validator.find_stake_account_address(program_id, accounts.lido.key, seed);
-        let stake_account = match stake_accounts.next() {
-            None => {
-                msg!(
-                    "Not enough stake accounts provided, got {} but expected {}.",
-                    accounts.stake_accounts.len(),
-                    end - begin,
-                );
-                msg!("Account at seed {} is missing.", seed);
-                return Err(LidoError::InvalidStakeAccount.into());
-            }
-            Some(account) if account.key != &stake_account_address => {
-                msg!(
-                    "Wrong stake account provided for seed {}: expected {} but got {}.",
-                    seed,
-                    stake_account_address,
-                    account.key,
-                );
-                return Err(LidoError::InvalidStakeAccount.into());
-            }
-            Some(account) => account,
-        };
-        let account_balance = Lamports(**stake_account.lamports.borrow());
-        msg!(
-            "Stake account at seed {} ({}) contains {}.",
+        let stake_account = &stake_accounts[i];
```

Comment:
If we zip with `stake_accounts` instead, that will allow the bounds check to be elided. Not sure if it makes things better or worse because I’m not sure if LLVM can see that those two iterators have the same length.

Reply:
I like it better with zip indeed :)